Docs: Grafana internal metrics update (#23448)
grafana/grafana: docs/sources/administration/metrics.md

weight = 8
+++

# Internal Grafana metrics

Grafana collects some metrics about itself internally. Grafana supports pushing metrics to Graphite or exposing them to be scraped by Prometheus.

For more information about configuration options related to Grafana metrics, refer to [metrics]({{< relref "../installation/configuration/#metrics" >}}) and [metrics.graphite]({{< relref "../installation/configuration/#metrics-graphite" >}}) in [Configuration]({{< relref "../installation/configuration.md" >}}).

## Available metrics

When enabled, Grafana exposes a number of metrics, including:

* Active Grafana instances
* Number of dashboards, users, and playlists
* http status codes
* Requests by routing group
* Grafana active alerts

> * Prometheus alerts

Author: Agred, fixed.

* Grafana performance

> ## View Grafana metrics in Prometheus

Reviewer: Maybe "View Grafana metrics stored in Prometheus" instead. "View Grafana metrics in Prometheus" makes me think this is about using the Prometheus web UI to render the metrics.

> ## View Grafana metrics in Prometheus

Author: What about Pull metrics from Grafana into Prometheus?

> ## View Grafana metrics in Prometheus

Reviewer: Sounds good!

These instructions assume you have already added Prometheus as a data source in Grafana.

> 1. Enable sending metrics to Prometheus. In your configuration file (`grafana.ini` or `custom.ini` depending on your operating system) remove the semicolon to enable the following configuration options: 

Reviewer: Something that is different for Prometheus compared to other data sources is that Prometheus collects (scrapes) metrics from applications/servers instead of applications/agents sending metrics to the data source. 

That's why you need to tell Prometheus (as you do on line 62) where it can find Grafana. While for Graphite you configure where to push metrics in Grafana.

Author: Do I need to make a change for this?

Reviewer: ```suggestion
1. Enable Prometheus to scrape metrics from Grafana. In your configuration file (`grafana.ini` or `custom.ini` depending on your operating system) remove the semicolon to enable the following configuration options: 
```

Reviewer: Sorry. Should have made a suggestion instead :)

   ```
   # Metrics available at HTTP API Url /metrics
   [metrics]
   # Disable / Enable internal metrics
   enabled           = true

   # Disable total stats (stat_totals_*) metrics to be generated
   disable_total_stats = false
   ```

2. (optional) If you want to require authorization to view the metrics endpoint, then uncomment and set the following options:

   ```
   basic_auth_username =
   basic_auth_password =
   ```

> 3. Restart Grafana. Grafana now sends metrics to http://localhost:3000/metrics.

Reviewer: ```suggestion
3. Restart Grafana. Grafana now exposes metrics at http://localhost:3000/metrics.
```

4. Add the job to your prometheus.yml file.
   Example:

   ```
   - job_name: 'grafana_metrics'

>       scrape_interval: 6s

Reviewer: ```suggestion
      scrape_interval: 15s
```

15s is the default so let's stick to that.
Kudos if you picked 6s because its a tenth of 60s :) But it's **very** uncommon to use that.

Author: Actually ended up with that because it was in a Getting started tutorial.

      scrape_timeout: 5s
  
      static_configs:
        - targets: ['localhost:3000']
   ```
5. Restart Prometheus. Your new job should appear on the Targets tab.
6. In Grafana, hover your mouse over the **Configuration** (gear) icon on the left sidebar and then click **Data Sources**.
7. Select the **Prometheus** data source.
8. On the Dashboards tab, **Import** the Grafana metrics dashboard. All scraped Grafana metrics are available in the dashboard.

## View Grafana metrics in Graphite

> These instructions assume you have already added Prometheus as a data source in Grafana.

Reviewer: ```suggestion
These instructions assume you have already added Graphite as a data source in Grafana.
```

Author: Heh. Copy/paste fail. Thank you!

1. Enable sending metrics to Graphite. In your configuration file (`grafana.ini` or `custom.ini` depending on your operating system) remove the semicolon to enable the following configuration options: 

   ```
   # Metrics available at HTTP API Url /metrics
   [metrics]
   # Disable / Enable internal metrics
   enabled           = true

   # Disable total stats (stat_totals_*) metrics to be generated
   disable_total_stats = false
   ```

> 1. (optional) If you want to require authorization to view the metrics endpoint, then uncomment and set the following options:

Reviewer: The basic_auth_username/password does not apply to Graphite. These settings are used to restrict access to the `http://localhost:3000/metrics' endpoint in Grafana that Prometheus scrapes metrics from.

Author: We should probably clarify that in the config file comments.
